sycl/source/detail/graph_impl.cpp

```diff
@@ -1027,141 +1042,297 @@ exec_graph_impl::~exec_graph_impl() {
+            UrEnqueueWaitList, nullptr);
+    return nullptr;
+  } else {
     auto NewEvent =
         std::make_shared<sycl::detail::event_impl>(Queue.shared_from_this());
     NewEvent->setContextImpl(Queue.getContextImplPtr());
     NewEvent->setStateIncomplete();
+    NewEvent->setSubmissionTime();
+    ur_event_handle_t UrEvent = nullptr;
+    Queue.getAdapter()
+        ->call<sycl::detail::UrApiKind::urEnqueueCommandBufferExp>(
+            Queue.getHandleRef(), CommandBuffer, UrEventHandles.size(),
+            UrEnqueueWaitList, &UrEvent);
+    NewEvent->setHandle(UrEvent);
+    NewEvent->setEventFromSubmittedExecCommandBuffer(true);
     return NewEvent;
-  });
-
-  sycl::detail::EventImplPtr NewEvent;
-  std::vector<sycl::detail::EventImplPtr> BackupCGDataMEvents;
-  if (MPartitions.size() > 1) {
-    BackupCGDataMEvents = CGData.MEvents;
-  }
-  for (uint32_t currentPartitionsNum = 0;
-       currentPartitionsNum < MPartitions.size(); currentPartitionsNum++) {
-    auto CurrentPartition = MPartitions[currentPartitionsNum];
-    // restore initial MEvents to add only needed additional depenencies
-    if (currentPartitionsNum > 0) {
-      CGData.MEvents = BackupCGDataMEvents;
-    }
-
-    for (auto const &DepPartition : CurrentPartition->MPredecessors) {
-      CGData.MEvents.push_back(PartitionsExecutionEvents[DepPartition]);
-    }
-
-    auto CommandBuffer = CurrentPartition->MCommandBuffers[Queue.get_device()];
-
-    if (CommandBuffer) {
-      for (std::vector<sycl::detail::EventImplPtr>::iterator It =
-               MExecutionEvents.begin();
-           It != MExecutionEvents.end();) {
-        EventImplPtr &Event = *It;
-        if (!Event->isCompleted()) {
-          const std::vector<EventImplPtr> &AttachedEventsList =
-              Event->getPostCompleteEvents();
-          CGData.MEvents.reserve(CGData.MEvents.size() +
-                                 AttachedEventsList.size() + 1);
-          CGData.MEvents.push_back(Event);
-          // Add events of the previous execution of all graph partitions.
-          CGData.MEvents.insert(CGData.MEvents.end(),
-                                AttachedEventsList.begin(),
-                                AttachedEventsList.end());
-          ++It;
-        } else {
-          // Remove completed events
-          It = MExecutionEvents.erase(It);
-        }
+  }
+}
+
+EventImplPtr
+exec_graph_impl::enqueuePartitions(sycl::detail::queue_impl &Queue,
+                                   sycl::detail::CG::StorageInitHelper &CGData,
+                                   bool IsCGDataSafeForSchedulerBypass,
+                                   bool EventNeeded) {
+
+  // If EventNeeded is true, this vector is used to keep track of dependencies
+  // for the returned event. This is used when the graph has multiple end nodes
+  // which cannot be tracked with a single scheduler event.
+  std::vector<EventImplPtr> PostCompleteDependencies;
```

Comment:
I haven't changed `event_impl` to inherit from `std::enable_shared_from_this`, but once it's done, do we need raw or shared pointers here? If you can add a comment on that, that would be very helpful for the future refactoring.

Reply:
I added a TODO comment. I think that we can use raw pointers there as long as the lifetime of the event is extended by the attachEventToComplete() implementation.